HIRS_Utils/src/main/java/hirs/validation/SupplyChainCredentialValidator.java

```diff
@@ -500,6 +575,115 @@ static AppraisalStatus validatePlatformCredentialAttributesV2p0(
+     * validated.
+     * @param origPcComponents The component identifier list associated with the
+     * base cert for this specific chain
+     * @return Appraisal Status of delta being validated.
+     */
+    static AppraisalStatus validateDeltaAttributesChainV2p0(
+            final DeviceInfoReport deviceInfoReport,
+            final List<PlatformCredential> leafChain,
+            final List<ComponentIdentifier> origPcComponents) {
+        boolean fieldValidation = true;
+        StringBuilder resultMessage = new StringBuilder();
+        List<ComponentIdentifier> validOrigPcComponents = origPcComponents.stream()
+                .filter(identifier -> identifier.getComponentManufacturer() != null
+                        && identifier.getComponentModel() != null)
+                .collect(Collectors.toList());
+
+        // map the components throughout the chain
+        Map<String, ComponentIdentifier> chainCiMapping = new HashMap<>();
+        List<ComponentIdentifier> deltaBuildList = new LinkedList<>(validOrigPcComponents);
+        deltaBuildList.stream().forEach((ci) -> {
+            chainCiMapping.put(ci.getComponentSerial().toString(), ci);
+        });
+
+        String ciSerial;
+        // go through the leaf and check the changes against the valid components
+        // forget modifying validOrigPcComponents
+        for (PlatformCredential delta : leafChain) {
+            for (ComponentIdentifier ci : delta.getComponentIdentifiers()) {
+                if (ci.isVersion2()) {
+                    ciSerial = ci.getComponentSerial().toString();
+                    ComponentIdentifierV2 ciV2 = (ComponentIdentifierV2) ci;
+
+                    if (ciV2.isModified())  {
+                        // this won't match
+                        // check it is there
+                        if (!chainCiMapping.containsKey(ciSerial)) {
+                            fieldValidation = false;
+                            resultMessage.append(String.format(
+                                    "%s attempted MODIFIED with no prior instance.%n",
+                                    ciSerial));
+                        } else {
+                            chainCiMapping.put(ci.getComponentSerial().toString(), ci);
+                        }
+                    } else if (ciV2.isRemoved()) {
+                        if (!chainCiMapping.containsKey(ciSerial)) {
+                            // error thrown, can't remove if it doesn't exist
+                            fieldValidation = false;
+                            resultMessage.append(String.format(
+                                    "%s attempted REMOVED with no prior instance.%n",
+                                    ciSerial));
+                        } else {
+                            chainCiMapping.remove(ci.getComponentSerial().toString());
+                        }
+                    } else {
```

Comment:
Need a "else if (ciV2.isAdded)" check? Could the status be EMPTY_STATUS here?

Reply:
I'm going to add it for completeness.  There shouldn't be the case in which the status is empty.  I did initially consider adding the method but didn't.